### PR TITLE
test: Add matchers for testing HeaderMaps.

### PR DESCRIPTION
**Pull request by mkbehr**
*Description*:

This adds:

- HeaderValueOf, for looking up a header and matching it against a string or any
  matcher.
- HttpStatusIs, for specifically testing the status header.
- IsSubsetOfHeaders, for testing that every desired header in a set of headers
  is present.

The older matchers HeaderHasValueRef and HeaderHasValue are now special cases of
HeaderValueOf and Pointee(HeaderValueOf), respectively.

*Risk Level*: Low (test only)
*Testing*: bazel test //test/...
*Docs Changes*: n/a
*Release Notes*: n/a

### CODE REVIEW

**Comment by mkbehr** [comment]
On second thought, I'm also going to throw these into a test here to make sure they work reasonably in that context.

**Comment by dnoe** [comment]
Nice, thanks for this. I'll wait to review once you fix-format, and assume that indicates your internal tests are all OK.

**Comment by mkbehr** [comment]
Added HttpStatusIs and HeaderValueOf to integration_test.

One thing I wasn't quite sure of was what namespace to put the new matchers in - Envoy::Http made sense to me, but some of the other matchers were just in Envoy.

File: test/mocks/http/mocks.cc [hunk 1]
```diff
@@ -160,5 +165,94 @@ MockInstance::MockInstance() {}
...
+    return Envoy::Http::HeaderMap::Iterate::Continue;
+  };
+  headers.iterate(get_headers_cb, &context);
+
+  if (values.empty()) {
+    *listener << "which has no '" << key_.get() << "' header";
+    return false;
+  } else if (values.size() > 1) {
+    *listener << "which has " << absl::StrCat(values.size()) << " '" << key_.get()
+              << "' headers, with values: " << absl::StrJoin(values, ", ");
```
Comment (dnoe) [inline test/mocks/http/mocks.cc]: It's not clear to me why `StrCat` is used here. Can't you just say `"which has " << values.size() << " '"`etc? The StrJoin makes sense.

File: test/mocks/http/mocks.cc [hunk 2]
```diff
@@ -160,5 +165,94 @@ MockInstance::MockInstance() {}
...
+    std::vector<std::pair<absl::string_view, absl::string_view>> arg_headers_vec;
+    headers.iterate(get_headers_cb, &arg_headers_vec);
+    std::vector<std::pair<absl::string_view, absl::string_view>> expected_headers_vec;
+    expected_headers_.iterate(get_headers_cb, &expected_headers_vec);
+
+    return ExplainMatchResult(IsSubsetOf(expected_headers_vec), arg_headers_vec, listener);
+  }
+
+  void DescribeTo(std::ostream* os) const override {
+    *os << "is subset of headers:\n" << expected_headers_;
```
Comment (dnoe) [inline test/mocks/http/mocks.cc]: nit: Is it more readable as "is a subset of headers" ?

File: test/mocks/http/mocks.h [hunk 3]
```diff
@@ -452,6 +455,54 @@ class MockInstance : public Instance {
...
+  return testing::MakeMatcher(
+      new HeaderValueOfMatcher(key, testing::SafeMatcherCast<absl::string_view>(matcher)));
+}
+
+template <typename T> testing::Matcher<const HeaderMap&> HeaderValueOf(std::string key, T matcher) {
+  return HeaderValueOf(LowerCaseString(key), matcher);
+}
+
+// Tests the provided Envoy HeaderMap for the provided HTTP status code.
+MATCHER_P(HttpStatusIs, expected_code, "") {
```
[the remark this answers is not in the document]
Comment (mkbehr) [reply test/mocks/http/mocks.h]: Done.

File: test/mocks/http/mocks.cc (hunk 1 above)
Comment (mkbehr) [reply test/mocks/http/mocks.cc]: Done.

File: test/mocks/http/mocks.cc (hunk 2 above)
Comment (mkbehr) [reply test/mocks/http/mocks.cc]: Done.

**Comment by dnoe** [comment]
@alyssawilk for senior maintainer review.

**Comment by alyssawilk** [comment]
I'm actually going to invoke the cross-company review - I normally wouldn't bother for tests, but I want to get a non-googler's opinion as to if the gmock version is easier to use and read than the original since we may be biased towards it.

Greg, given Dan's verified correctness can you just do a high level pass on what you think of the matchers?

**Comment by ggreenway** [comment]
I have mixed feelings. It looks like this is the "right" way to do things, but it also increases the barrier-to-entry for reading or writing tests. I now need to both understand how to use custom EXPECT_THAT predicates, and I need to know what these specific predicates do to read the test, whereas before I could look and see exactly which field it was digging out of the headers.

**Comment by dnoe** [comment]
I obviously come from the same biased position as @alyssawilk here, but I like the custom `EXPECT_THAT` matchers when they make the code in the test read more like a human language expression.

ie, `EXPECT_THAT(response->headers(), HttpStatusIs(200))` is more human readable than `EXPECT_STREQ("200", response->headers().Status()->value().c_str())`, and it's more important that the test be readable and understandable by humans than the compiler since it tells a story about the expectations.

While it hides some implementation, it's really just boilerplate of a form that the reader of the test doesn't need to concern themselves with: `->value().c_str()` is an implementation detail unimportant to most tests and adds visual clutter for the reader.

Would it help to create a top level  .md document (or under `test/`) listing available matchers and their purpose, linked from STYLE.md? This would certainly help with helping a test writer figure out what matchers exist.

For code reviewers/test readers, I'd say if it isn't very obvious what the matcher is doing from it's name then we should work to figure out a name that makes it very obvious.

**Comment by ggreenway** [comment]
@zuercher @ccaraman @junr03 @danielhochman @dio Any opinions on this?

**Comment by zuercher** [comment]
Looks good to me. I think as long as we're careful to make the matchers do the obvious thing, it should be ok.

**Comment by alyssawilk** [comment]
I think replacing the code is 95% of what we need for discoverability since most people copy-paste-edit anyway :-)    

We have a README.md under test/integration for how to write integration tests.  How about we land this as is but as a follow up we start a README under test/ for how to write Envoy tests, where we link to gtest/gmock, note the matchers, and link to the integration README?  As we add more tools we can extend that doc.

**Pull request merged**